### PR TITLE
Recursively check reachability when removing asset graphs from bundles in deduplication

### PR DESCRIPTION
**Pull request by wbinnssmith**
Supersedes/Closes #4931

With deduplication occurring between sibling bundles, previously it was possible to remove an asset graph from one sibling and remove a parent asset graph from the other, leaving an incomplete subgraph available in a bundle's ancestry. This adds an option to check assets as they're recursively removed to ensure they remain reachable from the bundle they're removed from.

Test Plan: This was tested in an app that reproduced this problem.

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.33s | +15.00ms |
| Cached | 2.16s | -84.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/index.html | 701.00b | +0.00b | 685.00ms | +61.00ms ⚠️ |
| dist/modern/index.html | 701.00b | +0.00b | 682.00ms | +60.00ms ⚠️ |
| dist/modern/index.e480862d.css | 77.00b | +0.00b | 1.30s | -103.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 231.00ms | +79.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 230.00ms | +78.00ms ⚠️ |
| dist/legacy/index.42516b4a.js | 1.10kb | +0.00b | 388.00ms | +29.00ms ⚠️ |
| dist/modern/index.8a1ced31.js | 1.10kb | +0.00b | 387.00ms | +29.00ms ⚠️ |
| dist/legacy/index.html | 701.00b | +0.00b | 395.00ms | +35.00ms ⚠️ |
| dist/modern/index.html | 701.00b | +0.00b | 387.00ms | +30.00ms ⚠️ |
| dist/legacy/index.0e158b6b.css | 77.00b | +0.00b | 394.00ms | +35.00ms ⚠️ |
| dist/modern/index.e480862d.css | 77.00b | +0.00b | 387.00ms | +29.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 27.63s | -109.00ms |
| Cached | 2.32s | +9.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 478.42kb | +0.00b | 400.00ms | +84.00ms ⚠️ |
| dist/PermalinkedComment.c85d137d.js | 4.22kb | +0.00b | 390.00ms | +75.00ms ⚠️ |
| dist/UserProfile.30e15baf.js | 1.62kb | +0.00b | 390.00ms | +75.00ms ⚠️ |
| dist/NotFound.5aa6b041.js | 387.00b | +0.00b | 390.00ms | +75.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 260.00ms | +18.00ms ⚠️ |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.79m | -2.45s |
| Cached | 5.04s | -50.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/EmojiPickerComponent.bab709db.js | 97.89kb | +0.00b | 12.19s | -1.43m 🚀 |
| dist/Toolbar.33a83ae3.js | 92.34kb | +0.00b | 55.02s | +2.78s ⚠️ |
| dist/media-viewer.5da4d79c.js | 72.01kb | +0.00b | 15.72s | +2.85s ⚠️ |
| dist/card.ce67dc35.js | 55.09kb | +0.00b | 12.91s | +2.81s ⚠️ |
| dist/Modal.e3bd77d0.js | 41.40kb | +0.00b | 5.52s | -385.00ms 🚀 |
| dist/component.59edff1f.js | 30.78kb | +0.00b | 2.34s | -564.00ms 🚀 |
| dist/component.37867f50.js | 22.51kb | +0.00b | 5.52s | -385.00ms 🚀 |
| dist/DatePicker.cbe44a9b.js | 21.17kb | +0.00b | 10.00s | -44.39s 🚀 |
| dist/EmojiPickerComponent.898d8be6.js | 18.23kb | +0.00b | 10.00s | -44.38s 🚀 |
| dist/dropzone.4407924e.js | 17.27kb | +0.00b | 51.39s | +2.95s ⚠️ |
| dist/smartMediaEditor.741b3eeb.js | 16.86kb | +0.00b | 53.89s | +2.87s ⚠️ |
| dist/js.48be70a4.js | 16.40kb | +0.00b | 5.52s | -385.00ms 🚀 |
| dist/ui.cee430a3.js | 13.97kb | +0.00b | 12.91s | +2.85s ⚠️ |
| dist/workerHasher.65b8a1b7.js | 11.78kb | +0.00b | 53.36s | +44.92s ⚠️ |
| dist/EmojiTypeAheadComponent.9af6e5fc.js | 8.67kb | +0.00b | 53.36s | +44.92s ⚠️ |
| dist/component.11476a92.js | 6.05kb | +0.00b | 5.52s | -385.00ms 🚀 |
| dist/card.4ac5c1b2.js | 5.61kb | +0.00b | 14.47s | +4.41s ⚠️ |
| dist/media-viewer.bc661bce.js | 3.92kb | +0.00b | 15.71s | +2.87s ⚠️ |
| dist/EmojiPickerComponent.624bf3f4.js | 3.58kb | +0.00b | 10.00s | -44.38s 🚀 |
| dist/index.393afe59.css | 3.46kb | +0.00b | 1.60m | +42.67s ⚠️ |
| dist/png-chunks-extract.8d15a914.js | 3.41kb | +0.00b | 5.52s | -385.00ms 🚀 |
| dist/dropzone.2f1a1e17.js | 3.23kb | +0.00b | 47.98s | +21.64s ⚠️ |
| dist/Modal.2835744c.js | 3.08kb | +0.00b | 4.54s | +270.00ms ⚠️ |
| dist/16.d1b8e6f7.js | 2.38kb | +0.00b | 2.34s | -564.00ms 🚀 |
| dist/EmojiUploadComponent.88a444df.js | 2.18kb | +0.00b | 53.36s | +44.92s ⚠️ |
| dist/card.cadccd96.js | 2.03kb | +0.00b | 29.38s | +16.50s ⚠️ |
| dist/workerHasher.9bb425b1.js | 1.54kb | +0.00b | 53.36s | +44.92s ⚠️ |
| dist/16.cff410ff.js | 1.38kb | +0.00b | 5.75s | -338.00ms 🚀 |
| dist/16.6603f2f3.js | 1.33kb | +0.00b | 5.75s | -337.00ms 🚀 |
| dist/16.a4b392c2.js | 1.32kb | +0.00b | 2.31s | -587.00ms 🚀 |
| dist/16.e60b4d77.js | 1.21kb | +0.00b | 2.31s | -585.00ms 🚀 |
| dist/media-card-analytics-error-boundary.6a25b970.js | 1.15kb | +0.00b | 29.52s | +3.18s ⚠️ |
| dist/16.2096b9f3.js | 1.14kb | +0.00b | 2.31s | -587.00ms 🚀 |
| dist/16.f040d62d.js | 1.14kb | +0.00b | 5.75s | -338.00ms 🚀 |
| dist/16.ecf8a702.js | 1.13kb | +0.00b | 8.10s | +685.00ms ⚠️ |
| dist/16.1b342686.js | 1.13kb | +0.00b | 5.75s | -338.00ms 🚀 |
| dist/16.2e7f52c6.js | 1.13kb | +0.00b | 5.75s | -338.00ms 🚀 |
| dist/quote.f4e09143.js | 1.02kb | +0.00b | 6.90s | -353.00ms 🚀 |
| dist/panel-warning.68b04096.js | 1008.00b | +0.00b | 6.90s | -353.00ms 🚀 |
| dist/panel.9dbc048e.js | 897.00b | +0.00b | 6.90s | -353.00ms 🚀 |
| dist/panel-success.7eb0809a.js | 845.00b | +0.00b | 6.90s | -353.00ms 🚀 |
| dist/panel-note.ec8386b7.js | 841.00b | +0.00b | 6.90s | -353.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.553055f6.js | 837.00b | +0.00b | 50.09s | +23.74s ⚠️ |
| dist/media-card-analytics-error-boundary.964206cc.js | 833.00b | +0.00b | 12.91s | +2.85s ⚠️ |
| dist/media-viewer-analytics-error-boundary.bbe40a02.js | 757.00b | +0.00b | 15.71s | +2.87s ⚠️ |
| dist/EmojiPickerComponent.a968dafe.js | 535.00b | +0.00b | 10.00s | +1.57s ⚠️ |
| dist/simpleHasher.f8e69464.js | 528.00b | +0.00b | 12.91s | +2.85s ⚠️ |
| dist/index.html | 119.00b | +0.00b | 739.00ms | +67.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.d95067b9.js | 2.73mb | +6.00b ⚠️ | 514.00ms | -46.00ms 🚀 |
| dist/pdfRenderer.cd742f5e.js | 1.11mb | -4.00b 🚀 | 510.00ms | -38.00ms 🚀 |
| dist/editorView.ae59b49f.js | 585.98kb | +0.00b | 511.00ms | -36.00ms 🚀 |
| dist/popup.4a66e15e.js | 170.62kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/EmojiPickerComponent.bab709db.js | 97.89kb | +0.00b | 511.00ms | -41.00ms 🚀 |
| dist/Toolbar.33a83ae3.js | 92.34kb | +0.00b | 511.00ms | -36.00ms 🚀 |
| dist/media-viewer.5da4d79c.js | 72.01kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/card.ce67dc35.js | 55.09kb | +0.00b | 510.00ms | -41.00ms 🚀 |
| dist/card.1007a95f.js | 52.86kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/Modal.e3bd77d0.js | 41.40kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/component.59edff1f.js | 30.78kb | +0.00b | 514.00ms | -46.00ms 🚀 |
| dist/component.37867f50.js | 22.51kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/DatePicker.cbe44a9b.js | 21.17kb | +0.00b | 511.00ms | -41.00ms 🚀 |
| dist/EmojiPickerComponent.898d8be6.js | 18.23kb | +0.00b | 511.00ms | -41.00ms 🚀 |
| dist/dropzone.4407924e.js | 17.27kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/smartMediaEditor.741b3eeb.js | 16.86kb | +0.00b | 511.00ms | -37.00ms 🚀 |
| dist/js.48be70a4.js | 16.40kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/ui.cee430a3.js | 13.97kb | +0.00b | 511.00ms | -40.00ms 🚀 |
| dist/workerHasher.65b8a1b7.js | 11.78kb | +0.00b | 511.00ms | -40.00ms 🚀 |
| dist/EmojiTypeAheadComponent.9af6e5fc.js | 8.67kb | +0.00b | 511.00ms | -40.00ms 🚀 |
| dist/component.11476a92.js | 6.05kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/card.4ac5c1b2.js | 5.61kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/media-viewer.bc661bce.js | 3.92kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/EmojiPickerComponent.624bf3f4.js | 3.58kb | +0.00b | 511.00ms | -41.00ms 🚀 |
| dist/index.393afe59.css | 3.46kb | +0.00b | 511.00ms | -36.00ms 🚀 |
| dist/png-chunks-extract.8d15a914.js | 3.41kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/dropzone.2f1a1e17.js | 3.23kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/Modal.2835744c.js | 3.08kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/clipboard.0244488f.js | 2.75kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/16.d1b8e6f7.js | 2.38kb | +0.00b | 514.00ms | -45.00ms 🚀 |
| dist/EmojiUploadComponent.88a444df.js | 2.18kb | +0.00b | 511.00ms | -40.00ms 🚀 |
| dist/card.cadccd96.js | 2.03kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/ResourcedEmojiComponent.d9a49606.js | 2.00kb | +0.00b | 514.00ms | -46.00ms 🚀 |
| dist/date.d49e64f8.js | 1.73kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/16.f40c1329.js | 1.73kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/images.80270b1f.js | 1.67kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/16.86a8c7d4.js | 1.66kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/feedback.8b7c7928.js | 1.64kb | +0.00b | 512.00ms | -46.00ms 🚀 |
| dist/browser.16c438d4.js | 1.61kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/workerHasher.9bb425b1.js | 1.54kb | +0.00b | 511.00ms | -40.00ms 🚀 |
| dist/workerHasher.f66f00c7.js | 1.54kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/list-number.4632b459.js | 1.46kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/status.cccf04be.js | 1.46kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/16.cff410ff.js | 1.38kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/code.af9ca2b3.js | 1.38kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/16.c1ed07f8.js | 1.33kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/16.6603f2f3.js | 1.33kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/16.a4b392c2.js | 1.32kb | +0.00b | 514.00ms | -46.00ms 🚀 |
| dist/16.249c644a.js | 1.31kb | +0.00b | 513.00ms | -47.00ms 🚀 |
| dist/link.d12cc130.js | 1.30kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/heading6.4c4017db.js | 1.30kb | +0.00b | 512.00ms | -46.00ms 🚀 |
| dist/heading3.93606b9c.js | 1.29kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/16.ab5c0088.js | 1.27kb | +0.00b | 513.00ms | -47.00ms 🚀 |
| dist/emoji.e603e70e.js | 1.23kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/16.afee1494.js | 1.23kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/16.e60b4d77.js | 1.21kb | +0.00b | 514.00ms | -46.00ms 🚀 |
| dist/16.7c7a40c9.js | 1.19kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/16.06d5a67b.js | 1.18kb | +0.00b | 513.00ms | -47.00ms 🚀 |
| dist/heading5.dde15cec.js | 1.17kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/expand.6889aead.js | 1.16kb | +0.00b | 511.00ms | -41.00ms 🚀 |
| dist/media-card-analytics-error-boundary.6a25b970.js | 1.15kb | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/16.2096b9f3.js | 1.14kb | +0.00b | 514.00ms | -46.00ms 🚀 |
| dist/16.f040d62d.js | 1.14kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/16.ecf8a702.js | 1.13kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/16.1b342686.js | 1.13kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/16.2e7f52c6.js | 1.13kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/16.da43aa07.js | 1.13kb | +0.00b | 513.00ms | -46.00ms 🚀 |
| dist/heading2.e219309a.js | 1.11kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/16.dd07505e.js | 1.10kb | +0.00b | 514.00ms | -46.00ms 🚀 |
| dist/component.40194190.js | 1.07kb | +0.00b | 514.00ms | -46.00ms 🚀 |
| dist/mention.b1828447.js | 1.07kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/heading4.5316f6f2.js | 1.06kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/16.317b9164.js | 1.05kb | +0.00b | 513.00ms | -47.00ms 🚀 |
| dist/layout.ac6734f9.js | 1.05kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/divider.d2be55a1.js | 1.03kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/quote.f4e09143.js | 1.02kb | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/action.cb287eb0.js | 1.00kb | +0.00b | 513.00ms | -47.00ms 🚀 |
| dist/decision.91f3af14.js | 1010.00b | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/panel-warning.68b04096.js | 1008.00b | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/list.057cb664.js | 978.00b | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/heading1.fa66b375.js | 976.00b | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/panel-error.c4b05f6f.js | 906.00b | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/panel.9dbc048e.js | 897.00b | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/table.b35545be.js | 889.00b | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/panel-success.7eb0809a.js | 845.00b | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/panel-note.ec8386b7.js | 841.00b | +0.00b | 512.00ms | -47.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.553055f6.js | 837.00b | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/media-card-analytics-error-boundary.964206cc.js | 833.00b | +0.00b | 510.00ms | -41.00ms 🚀 |
| dist/media-card-analytics-error-boundary.57503222.js | 833.00b | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/media-viewer-analytics-error-boundary.bbe40a02.js | 757.00b | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/EmojiPickerComponent.a968dafe.js | 535.00b | +0.00b | 511.00ms | -41.00ms 🚀 |
| dist/simpleHasher.f8e69464.js | 528.00b | +0.00b | 511.00ms | -40.00ms 🚀 |
| dist/simpleHasher.32d9352e.js | 528.00b | +0.00b | 510.00ms | -38.00ms 🚀 |
| dist/index.html | 119.00b | +0.00b | 516.00ms | -48.00ms 🚀 |

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/272601315641655816)**

File: packages/core/types/index.js [hunk 1]
```diff
@@ -913,7 +913,11 @@ export interface MutableBundleGraph extends BundleGraph<Bundle> {
   getParentBundlesOfBundleGroup(BundleGroup): Array<Bundle>;
   getTotalSize(Asset): number;
   /** Remove all "contains" edges from the bundle to the nodes in the asset's subgraph. */
-  removeAssetGraphFromBundle(Asset, Bundle): void;
+  removeAssetGraphFromBundle(
+    Asset,
+    Bundle,
+    ?$Shape<{|checkReachability: boolean|}>,
```
Comment (mischnic) [inline packages/core/types/index.js:919]:
```suggestion
    ?{checkReachability: boolean, ...}>,
```
I think that is equivalent?

Comment (wbinnssmith) [reply packages/core/types/index.js:919]: An empty object is incompatible with that unfortunately: https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAKA3gXwFxhYA+AHgPwEB2ArgLYBGApgE5E4CUQA

Comment (mischnic) [reply packages/core/types/index.js:919]: 🙈  Unsealed objects should just be removed

**Comment by devongovett** [comment]
Why is checkReachability an option as opposed to always on?

**Comment by wbinnssmith** [comment]
@devongovett 

> Why is checkReachability an option as opposed to always on?

It's a performance optimization. When we move asset graphs into common shared bundles, the entire subgraph is copied regardless of its reachability from the current bundle. Once we do that, we remove the graph from the original bundle. Since it's not possible for the subgraph to be incomplete in this new sibling bundle, we don't need to check. 

Admittedly, since it's now in a sibling, this is probably a pretty fast check bound by the number of siblings the bundle has rather than its entire ancestry.

File: packages/bundlers/default/src/DefaultBundler.js [hunk 2]
```diff
@@ -353,7 +353,6 @@ export default (new Bundler({
       .filter(bundle => bundle.size >= config.minBundleSize)
       .sort((a, b) => b.size - a.size);
 
-    let sharedBundles = [];
```
Comment (wbinnssmith) [inline packages/bundlers/default/src/DefaultBundler.js:356]: this was unused

**Pull request merged**